### DISCUSSION TITLE
Improve Producers configuration management

### DISCUSSION THREAD

**Discussion by iMacTia** [flipp-oss/deimos#146] [Ideas]
According to the [Configuration documentation](https://github.com/flipp-oss/deimos/blob/master/docs/CONFIGURATION.md#defining-producers), you define producers by using `producer do ... end` blocks in the `Deimos.configure` block.
This is handy at the beginning, but has some drawbacks once the project becomes bigger:
1. Producers need to be already loaded when the `Deimos.configure` block is executed in order for the configuration to be applied. This means the configuration is not applied correctly (e.g. `Myproducer.topic` is `nil`) if the producer is auto-loaded in a Rails application.
2. As the number of producers grows (we're planning to have one producer per event, which means multiple producers per topic), the `Deimos.configure` block grows as well and may become really big!

There might be some ways to address both these issues that I'm unaware of, but if that's not the case then here is a couple of ideas on how this could be addressed:

1. Allow producers to encapsulate their configuration: `topic` can already be set on the producer itself, but the same doesn't apply to other producer's configs that currently needs to go in the `Deimos.configure` block.
2. Allow producers to self-register: it would be great if the producer itself could register with Deimos, avoiding the need to have them all in the `Deimos.configure` block.

**Comment by dorner** [comment]
We have some apps with really big configs. We actually started with the topic etc. living with the producer, but more often we wanted to see all the consumers / producers in the app (e.g. to check the group IDs, topic names etc.) rather than each one individually. We generally have a separate initializer for Deimos so everything is pretty self-contained. I can definitely see benefits either way.

I did see the autoload issue in the past, but not often. That can maybe be mitigated by having the producer check the Deimos configuration when it's loading itself (I'm not sure if Ruby actually has a hook for this).

One thing I'm a bit concerned about is that you wouldn't want to allow consumers to register themselves (since consumers need to be fully defined before the listener is started) so you end up working with producers and consumers in different ways, which might be confusing.

I'm definitely open to a PR that allows both ways of operating. A way of checking/updating Deimos's configuration repository might be able to fit both of them.

**Reply by iMacTia** [reply]
That's actually a good point about having a holistic view of all your producers, you'd surely lose that without it.
But I also see that working with a proper folder structure where producers are all in the same place and organised in folders for grouping would probably solve the issue.

I'll see if there's a way to allow both as I can see people might prefer one or another based on their individual needs.

Point taken on the consumers, I see the problem there and will give that a thought as well 🤔
